lib/subroutine/fields.rb

```diff
@@ -159,88 +159,101 @@ def setup_fields(inputs = {})
       if ::Subroutine::Fields.action_controller_params_loaded? && inputs.is_a?(::ActionController::Parameters)
         inputs = inputs.to_unsafe_h if inputs.respond_to?(:to_unsafe_h)
       end
-      @provided_fields = {}.with_indifferent_access
-      param_groups[:original] = inputs.with_indifferent_access
+      inputs.each_pair do |k, v|
+        set_param_group_value(:original, k, v)
+      end
       mass_assign_initial_params
     end
 
+    def include_defaults_in_params?
+      self.class.include_defaults_in_params?
+    end
+
+    def param_cache
+      @param_cache ||= {}
+    end
+
     def param_groups
       @param_groups ||= Hash.new { |h, k| h[k] = {}.with_indifferent_access }
     end
 
     def get_param_group(name)
-      param_groups[name.to_sym]
+      name = name.to_sym
+      return param_groups[name] if param_groups.key?(name)
+
+      param_groups[name] = yield if block_given?
+      param_groups[name]
     end
 
-    def original_params
-      get_param_group(:original)
+    def set_param_group_value(name, key, value)
+      get_param_group(name)[key.to_sym] = value
     end
 
-    def ungrouped_params
-      get_param_group(:ungrouped)
+    def original_params
+      get_param_group(:original)
     end
 
-    def all_params
-      get_param_group(:all)
+    def all_provided_params
+      get_param_group(:provided)
     end
-    alias params all_params
+    alias provided_params all_provided_params
 
     def all_default_params
       get_param_group(:default)
     end
     alias defaults all_default_params
+    alias default_params all_default_params
 
-    def all_params_with_defaults
-      all_default_params.merge(all_params)
-    end
-    alias params_with_defaults all_params_with_defaults
+    def all_params
+      return all_params_with_default_params if include_defaults_in_params?
 
-    def ungrouped_defaults
-      default_params.slice(*ungrouped_fields.keys)
+      all_provided_params
```

Comment:
In the context of shapes and equality, have been wondering if `all_params` should return a hash of all fields (defaulting to nil or default_param if opted in), regardless of whether provided or not. Or do you think, this behavior should be separate - having a different method (other than `all_params`).

Maybe it's just something to implement In Glib.shape, where we currently do the following
```ruby
    def dump
      as_json
    end

    def as_json(*)
      ::Glib.attributes_sanitizer.sanitize_value(all_params_with_defaults)
    end
```
and kinda feel dumping the shape should represent all fields.